File: backend/src/test/java/org/globe42/web/security/AuthenticationFilterTest.java
package org.globe42.web.security;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

import java.io.IOException;
import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.Cookie;
import javax.servlet.http.HttpServletRequest;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.JwtException;
import io.jsonwebtoken.impl.DefaultClaims;
import org.globe42.dao.UserDao;
import org.globe42.test.BaseTest;
import org.junit.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.Spy;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.mock.web.MockHttpServletRequest;
import org.springframework.mock.web.MockHttpServletResponse;

/**
 * Unit tests for {@link AuthenticationFilter}
 * @author JB Nizet
 */
public class AuthenticationFilterTest extends BaseTest {

    @Spy
    private CurrentUser currentUser = new CurrentUser();

    @Mock
    private JwtHelper mockJwtHelper;

    @Mock
    private UserDao mockUserDao;

    @InjectMocks
    private AuthenticationFilter filter;

    @Mock
    private FilterChain mockFilterChain;

    @Test
    public void shouldRejectIfNoHeaderAndNoCookieForApiRequest() throws IOException, ServletException {
        MockHttpServletRequest request = new MockHttpServletRequest();
        request.setRequestURI("/api/persons");

        shouldRejectWithUnauthorized(request);
    }

    @Test
    public void shouldAcceptIfNoHeaderAndNoCookieForApiAuthenticationRequest() throws IOException, ServletException {
        MockHttpServletRequest request = new MockHttpServletRequest();
        request.setRequestURI("/api/authentication");

        shouldAccept(request);
    }

    @Test
    public void shouldRejectIfNoHeaderAndNoCookieForActuatorRequest() throws IOException, ServletException {
        MockHttpServletRequest request = new MockHttpServletRequest();
        request.setRequestURI("/actuator");

        shouldRejectWithUnauthorized(request);
    }

    @Test
    public void shouldAcceptIfNoHeaderAndNoCookieForActuatorHealthRequest() throws IOException, ServletException {
        MockHttpServletRequest request = new MockHttpServletRequest();
        request.setRequestURI("/actuator/health");

        shouldAccept(request);
    }

    @Test
    public void shouldRejectIfHeaderWithNoBearerPrefix() throws IOException, ServletException {
        MockHttpServletRequest request = new MockHttpServletRequest();
        request.setRequestURI("/api/persons");
        request.addHeader(HttpHeaders.AUTHORIZATION, "hello world");

        shouldRejectWithUnauthorized(request);
    }

    @Test
    public void shouldRejectIfHeaderWithBadToken() throws IOException, ServletException {
        MockHttpServletRequest request = new MockHttpServletRequest();
        request.setRequestURI("/api/persons");
        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer hello");

        when(mockJwtHelper.extractClaims("hello")).thenThrow(new JwtException("invalid"));

        shouldRejectWithUnauthorized(request);
    }

    @Test
    public void shouldRejectIfHeaderWithTokenWithBadUserId() throws IOException, ServletException {
        MockHttpServletRequest request = new MockHttpServletRequest();
        request.setRequestURI("/api/persons");
        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer hello");

        Claims claims = new DefaultClaims();
        claims.setSubject("abcd");
        when(mockJwtHelper.extractClaims("hello")).thenReturn(claims);

        shouldRejectWithUnauthorized(request);
    }

    @Test
    public void shouldRejectIfUserDoesNotExist() throws IOException, ServletException {
        MockHttpServletRequest request = new MockHttpServletRequest();
        request.setRequestURI("/api/persons");
        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer hello");

        Claims claims = new DefaultClaims();
        claims.setSubject("1234");
        when(mockJwtHelper.extractClaims("hello")).thenReturn(claims);
        when(mockUserDao.existsNotDeletedById(1234L)).thenReturn(false);

        shouldRejectWithUnauthorized(request);
    }

    @Test
    public void shouldSetCurrentUserIfValidTokenInHeader() throws IOException, ServletException {
        MockHttpServletRequest request = new MockHttpServletRequest();
        request.setRequestURI("/api/persons");
        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer hello");

        Claims claims = new DefaultClaims();
        claims.setSubject("1234");
        when(mockJwtHelper.extractClaims("hello")).thenReturn(claims);
        when(mockUserDao.existsNotDeletedById(1234L)).thenReturn(true);

        shouldAccept(request);
        assertThat(currentUser.getUserId()).isEqualTo(1234L);
    }

    @Test
    public void shouldSetCurrentUserIfValidTokenInCookie() throws IOException, ServletException {
        MockHttpServletRequest request = new MockHttpServletRequest();
        request.setRequestURI("/api/persons");
        request.setCookies(new Cookie("globe42_token", "hello"));

        Claims claims = new DefaultClaims();
        claims.setSubject("1234");
        when(mockJwtHelper.extractClaims("hello")).thenReturn(claims);
        when(mockUserDao.existsNotDeletedById(1234L)).thenReturn(true);

        shouldAccept(request);
        assertThat(currentUser.getUserId()).isEqualTo(1234L);
    }

    @Test
    public void shouldRejectIfNotAdminForProtectedActuatorRequest() throws IOException, ServletException {
        MockHttpServletRequest request = new MockHttpServletRequest();
        request.setRequestURI("/actuator/foo");
        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer hello");

        Claims claims = new DefaultClaims();
        claims.setSubject("1234");
        when(mockJwtHelper.extractClaims("hello")).thenReturn(claims);
        when(mockUserDao.existsNotDeletedById(1234L)).thenReturn(true);
        when(mockUserDao.existsNotDeletedAdminById(1234L)).thenReturn(false);

        shouldRejectWithForbidden(request);
    }

    private void shouldRejectWithUnauthorized(HttpServletRequest request) throws IOException, ServletException {
        MockHttpServletResponse response = new MockHttpServletResponse();
        filter.doFilter(request, response, mockFilterChain);

        verify(mockFilterChain, never()).doFilter(request, response);
        assertThat(response.getStatus()).isEqualTo(HttpStatus.UNAUTHORIZED.value());
    }

    private void shouldRejectWithForbidden(HttpServletRequest request) throws IOException, ServletException {
        MockHttpServletResponse response = new MockHttpServletResponse();
        filter.doFilter(request, response, mockFilterChain);

        verify(mockFilterChain, never()).doFilter(request, response);
        assertThat(response.getStatus()).isEqualTo(HttpStatus.FORBIDDEN.value());
    }

    private void shouldAccept(HttpServletRequest request) throws IOException, ServletException {
        MockHttpServletResponse response = new MockHttpServletResponse();
        filter.doFilter(request, response, mockFilterChain);

        verify(mockFilterChain).doFilter(request, response);
    }
}

File: backend/src/test/kotlin/org/globe42/web/security/AuthenticationFilterTest.kt
package org.globe42.web.security

import io.jsonwebtoken.JwtException
import io.jsonwebtoken.impl.DefaultClaims
import io.mockk.every
import io.mockk.mockk
import io.mockk.spyk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.globe42.dao.UserDao
import org.junit.jupiter.api.Test
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.mock.web.MockHttpServletRequest
import org.springframework.mock.web.MockHttpServletResponse
import javax.servlet.FilterChain
import javax.servlet.http.Cookie
import javax.servlet.http.HttpServletRequest

/**
 * Unit tests for [AuthenticationFilter]
 * @author JB Nizet
 */
class AuthenticationFilterTest {

    private val currentUser = spyk(CurrentUser())

    private val mockJwtHelper = mockk<JwtHelper>()

    private val mockUserDao = mockk<UserDao>()

    private val filter = AuthenticationFilter(mockJwtHelper, currentUser, mockUserDao)

    private val mockFilterChain = mockk<FilterChain>(relaxUnitFun = true)

    @Test
    fun `should reject if no header and no cookie for api request`() {
        val request = MockHttpServletRequest()
        request.requestURI = "/api/persons"

        shouldRejectWithUnauthorized(request)
    }

    @Test
    fun `should accept if no header and no cookie for api authentication request`() {
        val request = MockHttpServletRequest()
        request.requestURI = "/api/authentication"

        shouldAccept(request)
    }

    @Test
    fun `should reject if no header and no cookie for actuator request`() {
        val request = MockHttpServletRequest()
        request.requestURI = "/actuator"

        shouldRejectWithUnauthorized(request)
    }

    @Test
    fun `should accept if no header and no cookie for actuator health request`() {
        val request = MockHttpServletRequest()
        request.requestURI = "/actuator/health"

        shouldAccept(request)
    }

    @Test
    fun `should reject if header with no bearer prefix`() {
        val request = MockHttpServletRequest()
        request.requestURI = "/api/persons"
        request.addHeader(HttpHeaders.AUTHORIZATION, "hello world")

        shouldRejectWithUnauthorized(request)
    }

    @Test
    fun `should reject if header with bad token`() {
        val request = MockHttpServletRequest()
        request.requestURI = "/api/persons"
        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer hello")

        every { mockJwtHelper.extractClaims("hello") } throws JwtException("invalid")

        shouldRejectWithUnauthorized(request)
    }

    @Test
    fun `should reject if header with token with bad user id`() {
        val request = MockHttpServletRequest()
        request.requestURI = "/api/persons"
        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer hello")

        val claims = DefaultClaims()
        claims.subject = "abcd"
        every { mockJwtHelper.extractClaims("hello") } returns claims

        shouldRejectWithUnauthorized(request)
    }

    @Test
    fun `should reject if user does not exist`() {
        val request = MockHttpServletRequest()
        request.requestURI = "/api/persons"
        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer hello")

        val claims = DefaultClaims()
        claims.subject = "1234"
        every { mockJwtHelper.extractClaims("hello") } returns claims
        every { mockUserDao.existsNotDeletedById(1234L) } returns false

        shouldRejectWithUnauthorized(request)
    }

    @Test
    fun `should set current user if valid token in header`() {
        val request = MockHttpServletRequest()
        request.requestURI = "/api/persons"
        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer hello")

        val claims = DefaultClaims()
        claims.subject = "1234"
        every { mockJwtHelper.extractClaims("hello") } returns claims
        every { mockUserDao.existsNotDeletedById(1234L) } returns true

        shouldAccept(request)
        assertThat(currentUser.userId).isEqualTo(1234L)
    }

    @Test
    fun `should set current user if valid token in cookie`() {
        val request = MockHttpServletRequest()
        request.requestURI = "/api/persons"
        request.setCookies(Cookie("globe42_token", "hello"))

        val claims = DefaultClaims()
        claims.subject = "1234"
        every { mockJwtHelper.extractClaims("hello") } returns claims
        every { mockUserDao.existsNotDeletedById(1234L) } returns true

        shouldAccept(request)
        assertThat(currentUser.userId).isEqualTo(1234L)
    }

    @Test
    fun `should reject if not admin for protected actuator request`() {
        val request = MockHttpServletRequest()
        request.requestURI = "/actuator/foo"
        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer hello")

        val claims = DefaultClaims()
        claims.subject = "1234"
        every { mockJwtHelper.extractClaims("hello") } returns claims
        every { mockUserDao.existsNotDeletedById(1234L) } returns true
        every { mockUserDao.existsNotDeletedAdminById(1234L) } returns false

        shouldRejectWithForbidden(request)
    }

    private fun shouldRejectWithUnauthorized(request: HttpServletRequest) {
        val response = MockHttpServletResponse()
        filter.doFilter(request, response, mockFilterChain)

        verify(inverse = true) { mockFilterChain.doFilter(request, response) }
        assertThat(response.status).isEqualTo(HttpStatus.UNAUTHORIZED.value())
    }

    private fun shouldRejectWithForbidden(request: HttpServletRequest) {
        val response = MockHttpServletResponse()
        filter.doFilter(request, response, mockFilterChain)

        verify(inverse = true) { mockFilterChain.doFilter(request, response) }
        assertThat(response.status).isEqualTo(HttpStatus.FORBIDDEN.value())
    }

    private fun shouldAccept(request: HttpServletRequest) {
        val response = MockHttpServletResponse()
        filter.doFilter(request, response, mockFilterChain)

        verify { mockFilterChain.doFilter(request, response) }
    }
}
